File: leetcode/algorithms/202_happy_number/main.js
/**
 * Hash Set
 *
 * Complexities:
 *   N - `n`
 *   - Time Complexity: O(logᴺ)
 *   - Space Complexity: O(logᴺ)
 */
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  const set = {};

  while (n > 1) {
    let nString = n.toString();
    if (set[nString]) {
      return false;
    }
    set[nString] = true;
    n = nString.split("").reduce((total_, current_) => {
      return parseInt(total_) + Math.pow(parseInt(current_), 2);
    }, 0);
    if (n === 1) {
      return true;
    }
  }

  return n === 1;
};


// Solution
/**
 * Solution 1
 *
 * Floyd's Cycle-Finding Algorithm
 *
 * Complexities:
 *   N - `n`
 *   - Time Complexity: O(logᴺ)
 *   - Space Complexity: O(logᴺ)
 */
/**
 * @param {number} n
 * @return {number}
 */
function getNext1(n) {
  let totalSum = 0;

  while (n > 0) {
    const digit = n % 10;
    totalSum += digit * digit;
    n = Math.floor(n / 10);
  }

  return totalSum;
}

/**
 * @param {number} n
 * @return {boolean}
 */
var solution1 = function (n) {
  const seen = new Set();

  while (n !== 1 && !seen.has(n)) {
    seen.add(n);
    n = getNext1(n);
  }

  return n === 1;
};

/**
 * Solution 2
 *
 * Hash Set
 *
 * Complexities:
 *   N - `n`
 *   - Time Complexity: O(logᴺ)
 *   - Space Complexity: O(1)
 */
/**
 * @param {number} n
 * @return {number}
 */
function getNext2(n) {
  let totalSum = 0;

  while (n > 0) {
    const digit = n % 10;
    totalSum += digit * digit;
    n = Math.floor(n / 10);
  }

  return totalSum;
}

/**
 * @param {number} n
 * @return {boolean}
 */
var solution2 = function (n) {
  let slow = n;
  let fast = getNext2(n);

  while (fast !== 1 && slow !== fast) {
    slow = getNext2(slow);
    fast = getNext2(getNext2(fast));
  }

  return fast === 1;
};

File: leetcode/algorithms/202_happy_number/main.dart
class HappyNumber {
  /**
   * Floyd's Cycle-Finding Algorithm
   *
   * Complexities:
   *   N - `n`
   *   - Time Complexity: O(logᴺ)
   *   - Space Complexity: O(1)
   */
  int sum(int n) {
    int sum = 0;
    while (n > 0) {
      sum += (n % 10) * (n % 10);
      n = (n / 10).floor();
    }
    return sum;
  }

  bool isHappy(int n) {
    int fast = n;
    int slow = n;

    while (true) {
      slow = sum(slow);
      fast = sum(sum(fast));
      if (fast == 1) {
        break;
      }
      if (fast == slow) {
        return false;
      }
    }

    return true;
  }


  // Solution
  /**
   * Solution 1
   * 
   * Floyd's Cycle-Finding Algorithm
   *
   * Complexities:
   *   N - `n`
   *   - Time Complexity: O(logᴺ)
   *   - Space Complexity: O(1)
   */
  int _getNext1(int n) {
    int totalSum = 0;

    while (n > 0) {
      int digit = n % 10;
      n = n ~/ 10;
      totalSum += digit * digit;
    }

    return totalSum;
  }

  bool solution1(int n) {
    int slow = n;
    int fast = _getNext1(n);

    while (fast != 1 && slow != fast) {
      slow = _getNext1(slow);
      fast = _getNext1(_getNext1(fast));
    }

    return fast == 1;
  }

  /**
   * Solution 2
   * 
   * Hash Set
   *
   * Complexities:
   *   N - `n`
   *   - Time Complexity: O(logᴺ)
   *   - Space Complexity: O(logᴺ)
   */
  int _getNext2(int n) {
    int totalSum = 0;

    while (n > 0) {
      int digit = n % 10;
      n = n ~/ 10;
      totalSum += digit * digit;
    }

    return totalSum;
  }

  bool solution2(int n) {
    final Set<int> seen = {};

    while (n != 1 && !seen.contains(n)) {
      seen.add(n);
      n = _getNext2(n);
    }

    return n == 1;
  }
}
